### DISCUSSION TITLE
Novato usando git

### DISCUSSION THREAD

**Discussion by rubensilbano** [community/community#37655] [Other Feature Feedback, Questions, & Ideas] [answered]
Estoy por entrar a una jam usando UE4 con un colega. Y tengo las siguientes dudas:
¿En que momento se recomienda hacer push?
-Cuando termino una funcion libre de errores, al final de la jornada.
-Cada 10 minutos aprox.
-Es trivial, usar con la misma frecuencia que Ctrl+S

¿Ambos deberiamos trabajar en una misma rama o cada uno en una rama separada y juntarlas luego con merge?

Supongamos que por ejemplo:
Yo me encargo de los enemigos, y el contador de enemigos se guarda en el game mode.
Y mi colega se encarga de los HUD, y el de victoria lee las variables del game mode.
Y entre los dos terminamos haciendo dos versiones distintas de el mismo game mode, y ambos lo subimos.
¿Como se resuelve tal lio?

**Comment by DenverCoder1** [comment, marked as answer]
> ¿En que momento se recomienda hacer push?
> -Cuando termino una funcion libre de errores, al final de la jornada.

In my opinion, it is best to commit after completing a full piece of code. When you finish writing a function or making a fix, it is recommended to make a commit and push usually when it is fully working.

> ¿Ambos deberiamos trabajar en una misma rama o cada uno en una rama separada y juntarlas luego con merge?

If two people make changes to the same lines of code, it may cause a "merge conflict". This can happen whether you both work on the same branch or merge from separate branches.

It will appear in the file like this when it occurs:

```
some text before
<<<<<<< HEAD
This is one of the two
versions of the file that
was edited
=======
This is another of the two
versions of the file that was edited
>>>>>>> branch_to_merge
some text after
```

You should remove the `<<<<<<<` and `>>>>>>>` lines and choose which parts of each you want to keep. In the end, when the conflict is resolved and the file is back to normal, you can commit the merge.

```
some text before
This is final version
of the file that was edited
some text after
```

More info: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-using-the-command-line

There are also tools such as ones in VS Code that can make resolving merge conflicts easier.

**Reply by rubensilbano** [reply]
Thanks for your answer and time, and my apologies for my low skills on english and time delay on response. Actually i continue learn on deep the example Platformer of Unreal engine 4.

**Reply by airtower-luna** [reply]
> In my opinion, it is best to commit after completing a full piece of code. When you finish writing a function or making a fix, it is recommended to make a commit and push usually when it is fully working.

This is good advice. Building on it, try not to let your commits get *bigger* than that. For example if you have two bugs to fix, commit after fixing one, then approach the other. That isn't always practical (for example if the bugs are tightly connected), but it's a good rule of thumb. And ideally tests should pass for each commit.

**Reply by jfullstackdev** [reply]
yup, speaking of merge conflicts 😆 actually, I'm not using Git or GitHub Desktop, I just used GitHub Desktop in my assigned project just to clone it at first

what's happening is that, we are assigned per module, so no conflict will happen but I'm not ever synchronizing my local work from GitHub, when I'm developing locally, what I will do is to manually create branches on GitHub, then paste the files that have been changed, then will create a PR and merge it, so far, for many months, I'm not encountering any errors, for me, this way, I can feel that I committed and change only those necessary files

but when files that are being changed will be more than 3, that's I think is the time to use Git properly or many developers are assigned in one module
